src/nations/nations.cpp

```diff
@@ -211,10 +211,40 @@ bool identity_has_holder(sys::state const& state, dcon::national_identity_id ide
 }
 
 bool are_allied(sys::state& state, dcon::nation_id a, dcon::nation_id b) {
-	auto rel = state.world.get_diplomatic_relation_by_diplomatic_pair(a, b);
+	auto rel = get_diplomatic_relation(state, a, b);
 	return state.world.diplomatic_relation_get_are_allied(rel);
 }
 
+void get_allies(sys::state& state, dcon::nation_id n, std::vector<dcon::nation_id>& allies) {
+	for(auto dr : state.world.nation_get_diplomatic_relation_as_related_nations(n)) {
+		if(dr.get_are_allied()) {
+			auto other = dr.get_related_nations(0) != n ? dr.get_related_nations(0) : dr.get_related_nations(1);
+			allies.push_back(other);
+		}
+	}
+}
+
+// also returns true if one is the sphere leader of the other
+bool in_same_sphere(sys::state& state, dcon::nation_id a, dcon::nation_id b) {
+	auto a_sphere = fatten(state.world, a).get_in_sphere_of();
+	auto b_sphere = fatten(state.world, b).get_in_sphere_of();
+	if(a_sphere && a_sphere == b)
+		return true;
+	if(b_sphere && b_sphere == a)
+		return true;
+	if(a_sphere && b_sphere && a_sphere == b_sphere)
+		return true;
+	return false;
+}
+
+bool are_neighbors(sys::state& state, dcon::nation_id a, dcon::nation_id b) {
+	for(auto g : fatten(state.world, a).get_nation_adjacency()) {
```

Comment:
This can be reduced to `get_nation_adjacency_by_nation_adjacency_pair`.  This does a (fast) hash lookup using the combined indices and will return a valid handle (true if cast to bool) if they are adjacent and an invalid handle if they are not